In [1]:
import pandas as pd

In [2]:
url = ('/Users/javi/Desktop/Ironhack/Entregas de github/Personal-Project-at-Ironhack/2017Co2_Emissions+Europe.csv')
df= pd.read_csv(url)

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33138 entries, 0 to 33137
Data columns (total 6 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   country_long                   33138 non-null  object 
 1   name                           33138 non-null  object 
 2   primary_fuel                   33138 non-null  object 
 3   estimated_generation_gwh_2017  33138 non-null  float64
 4   CO2e/KWh                       32031 non-null  float64
 5   Total MtCO2                    32031 non-null  float64
dtypes: float64(3), object(3)
memory usage: 1.5+ MB


In [4]:
df['CO2e/KWh'].unique()

array([   4. ,   41. ,  610. ,  840. ,    nan,   12. , 1214.5,   11. ,
         38. ])

In [6]:
df.shape

(33138, 6)

In [8]:
df['primary_fuel'].value_counts()

primary_fuel
Solar         10665
Hydro          7156
Wind           5343
Gas            3947
Coal           2306
Oil            2263
Waste          1065
Nuclear         193
Geothermal      158
Other            42
Name: count, dtype: int64

In [10]:
df.head()

country_long                                              name primary_fuel  \
0  Afghanistan      Kajaki Hydroelectric Power Plant Afghanistan        Hydro   
1  Afghanistan                                      Kandahar DOG        Solar   
2  Afghanistan                                      Kandahar JOL        Solar   
3  Afghanistan     Mahipar Hydroelectric Power Plant Afghanistan        Hydro   
4  Afghanistan  Naghlu Dam Hydroelectric Power Plant Afghanistan        Hydro   

   estimated_generation_gwh_2017  CO2e/KWh  Total MtCO2  
0                         119.50       4.0       478.00  
1                          18.29      41.0       749.89  
2                          18.72      41.0       767.52  
3                         174.91       4.0       699.64  
4                         350.80       4.0      1403.20

country_long	name	primary_fuel	estimated_generation_gwh_2017	CO2e/KWh	Total MtCO2
0	Afghanistan	Kajaki Hydroelectric Power Plant Afghanistan	Hydro	119.50	4.0	478.00
1	Afghanistan	Kandahar DOG	Solar	18.29	41.0	749.89
2	Afghanistan	Kandahar JOL	Solar	18.72	41.0	767.52
3	Afghanistan	Mahipar Hydroelectric Power Plant Afghanistan	Hydro	174.91	4.0	699.64
4	Afghanistan	Naghlu Dam Hydroelectric Power Plant Afghanistan	Hydro	350.80	4.0	1403.20 


this a df i have imported from a csv and it´s showing you the head. As you can see it show countries. the name of the powerplant. which fuel_type it has (wind, coal, oil, gas, solar, hydro, etc.) and how many gwh it was estimated it has produced.  total production of co2 in metric tons using grams of CO2e/KWh as a reference (the conversion was made correctly showing the amopunt of Metric tons it has produced taking into account the gwh of that year production.

since they are more than 33000 powerplants (actually way more but we are just taking into account this datafram) it´s a bit hard to see the total results of everything. 

I want to have several grouped df that show different things:

1. is the highest produced energy produced in relation to the others by country (so total energy production of each country and show it as a percentage in addition to the total produced. 

2. is the country that has polluted the most generating this adding all of the 

In [ ]:
# Step 1: Group by country to calculate the total energy production and percentage of energy production for each country
total_energy_by_country = df.groupby('country_long')['estimated_generation_gwh_2017'].sum().reset_index()
total_energy_by_country['Percentage'] = (total_energy_by_country['estimated_generation_gwh_2017'] / total_energy_by_country['estimated_generation_gwh_2017'].sum()) * 100
total_energy_by_country = total_energy_by_country.sort_values(by='estimated_generation_gwh_2017', ascending=False)


In [32]:
print("Total Energy Production by Country:")
total_energy_by_country.head(10)

Total Energy Production by Country:


country_long  estimated_generation_gwh_2017  Percentage
23                      China                     6193117.75   26.085620
115  United States of America                     4653165.24   19.599289
37                     Europe                     4034112.86   16.991819
46                      India                     1434479.65    6.042077
52                      Japan                      998572.01    4.206019
19                     Canada                      623497.21    2.626191
98                South Korea                      548133.75    2.308758
13                     Brazil                      518866.87    2.185485
93               Saudi Arabia                      345439.30    1.455002
70                     Mexico                      327348.33    1.378802

In [48]:
# Step 4: Group by country and fuel type to calculate the percentage of energy production and its contribution to total pollution
energy_pollution_by_country_fuel = df.groupby(['country_long', 'primary_fuel']).agg({'estimated_generation_gwh_2017': 'sum', 'Total MtCO2': 'sum'}).reset_index()
energy_pollution_by_country_fuel['Percentage_Energy'] = (energy_pollution_by_country_fuel['estimated_generation_gwh_2017'] / energy_pollution_by_country_fuel['estimated_generation_gwh_2017'].sum()) * 100
energy_pollution_by_country_fuel['Percentage_Pollution'] = (energy_pollution_by_country_fuel['Total MtCO2'] / df['Total MtCO2'].sum()) * 100
energy_pollution_by_country_fuel = energy_pollution_by_country_fuel.sort_values(by='Total MtCO2', ascending=False)
print("Energy Production and Pollution Contribution by Country and Fuel Type:")
energy_pollution_by_country_fuel.head(50)

Energy Production and Pollution Contribution by Country and Fuel Type:


country_long primary_fuel  estimated_generation_gwh_2017  \
80                      China         Coal                     4656913.40   
386  United States of America         Coal                     1150861.91   
161                     India         Coal                     1079504.29   
387  United States of America          Gas                     1926994.19   
128                    Europe         Coal                      900192.15   
129                    Europe          Gas                     1114955.57   
190                     Japan         Coal                      272740.41   
331               South Korea         Coal                      250673.71   
323              South Africa         Coal                      229709.87   
19                  Australia         Coal                      164961.11   
168                 Indonesia         Coal                      159629.09   
191                     Japan          Gas                      241429.38   
313              Saudi Arabia          Gas                      211342.75   
172                      Iran          Gas                      201857.32   
239                    Mexico          Gas                      201404.55   
352                    Taiwan         Coal                      100338.58   
374                    Turkey         Coal                       99173.79   
202                Kazakhstan         Coal                       93250.53   
226                  Malaysia         Coal                       92749.98   
314              Saudi Arabia          Oil                      133781.32   
81                      China          Gas                      181152.39   
407                   Vietnam         Coal                       70339.02   
384      United Arab Emirates          Gas                      138219.44   
115                     Egypt          Gas                      134899.58   
332               South Korea          Gas                      130792.34   
64                     Canada         Coal                       64444.49   
363                  Thailand          Gas                      126640.09   
302               Philippines         Coal                       55020.26   
375                    Turkey          Gas                      101921.61   
195                     Japan          Oil                       73768.14   
162                     India          Gas                       78495.62   
45                     Brazil          Gas                       74396.62   
9                   Argentina          Gas                       73047.29   
65                     Canada          Gas                       69091.58   
227                  Malaysia          Gas                       66610.10   
362                  Thailand         Coal                       32679.03   
2                     Algeria          Gas                       63380.37   
408                   Vietnam          Gas                       60848.99   
238                    Mexico         Coal                       30561.89   
182                    Israel          Gas                       58565.77   
211                    Kuwait          Gas                       58477.28   
33                 Bangladesh          Gas                       57916.02   
20                  Australia          Gas                       55500.38   
309                     Qatar          Gas                       53987.72   
169                 Indonesia          Gas                       53375.59   
74                      Chile         Coal                       26698.18   
320                 Singapore          Gas                       52740.09   
133                    Europe          Oil                       38248.76   
8                   Argentina         Coal                       25175.76   
29                    Bahrain          Gas                       48993.90   

      Total MtCO2  Percentage_Energy  Percentage_Pollution  
80   5.655821e+09          1

In [42]:
total_co2_by_country = df.groupby('country_long')['Total MtCO2'].sum().reset_index()
total_co2_by_country['Percentage'] = (total_co2_by_country['Total MtCO2'] / total_co2_by_country['Total MtCO2'].sum()) * 100
total_co2_by_country = total_co2_by_country.sort_values(by='Total MtCO2', ascending=False)

# Store the results in a DataFrame
total_co2_by_country_df = pd.DataFrame(total_co2_by_country)

# Displaying the results
print("Total CO2 Emissions by Country (in Percentage):")
total_co2_by_country_df.head(10)

Total CO2 Emissions by Country (in Percentage):


country_long   Total MtCO2  Percentage
23                      China  5.779019e+09   36.366000
115  United States of America  2.609919e+09   16.423604
37                     Europe  1.825588e+09   11.487992
46                      India  1.369505e+09    8.617973
52                      Japan  5.452653e+08    3.431225
98                South Korea  3.941630e+08    2.480375
97               South Africa  2.795478e+08    1.759128
93               Saudi Arabia  2.413083e+08    1.518496
5                   Australia  2.366426e+08    1.489136
47                  Indonesia  2.275573e+08    1.431964

In [49]:
# Assuming df is your DataFrame

# Choose the country of interest
country_name = 'China'

# Filter the DataFrame by the country of interest
country_df = df[df['country_long'] == country_name]

# Group by fuel type to calculate the total energy production for each fuel type
energy_by_fuel_type = country_df.groupby('primary_fuel')['estimated_generation_gwh_2017'].sum().reset_index()
energy_by_fuel_type = energy_by_fuel_type.sort_values(by='estimated_generation_gwh_2017', ascending=False)

# Displaying the results
print(f"Fuel Types and Total Energy Production for {country_name}:")
print(energy_by_fuel_type.head())


Fuel Types and Total Energy Production for China:
  primary_fuel  estimated_generation_gwh_2017
0         Coal                     4656913.40
3        Hydro                      923375.89
4      Nuclear                      246593.44
1          Gas                      181152.39
7         Wind                       94613.98


In [58]:
# Assuming df is your DataFrame

# Select the top 5 countries with the highest total energy production
top_5_countries = total_energy_by_country.nlargest(5, 'estimated_generation_gwh_2017')

# Iterate over each of the top 5 countries
for country_name in top_5_countries['country_long']:
    # Filter the DataFrame by the country of interest
    country_df = df[df['country_long'] == country_name]

    # Calculate the total energy production for the country
    total_energy_country = country_df['estimated_generation_gwh_2017'].sum()

    # Group by fuel type to calculate the total energy production for each fuel type
    energy_by_fuel_type = country_df.groupby('primary_fuel')['estimated_generation_gwh_2017'].sum().reset_index()
    energy_by_fuel_type['Percentage'] = (energy_by_fuel_type['estimated_generation_gwh_2017'] / total_energy_country) * 100
    energy_by_fuel_type = energy_by_fuel_type.sort_values(by='Percentage', ascending=False)

    # Displaying the results for each country
    print(f"Fuel Types and Total Energy Production Percentage for {country_name}:")
    print(energy_by_fuel_type.head())
    print("\n")


Fuel Types and Total Energy Production Percentage for China:
  primary_fuel  estimated_generation_gwh_2017  Percentage
0         Coal                     4656913.40   75.194976
3        Hydro                      923375.89   14.909710
4      Nuclear                      246593.44    3.981733
1          Gas                      181152.39    2.925060
7         Wind                       94613.98    1.527728


Fuel Types and Total Energy Production Percentage for United States of America:
  primary_fuel  estimated_generation_gwh_2017  Percentage
1          Gas                     1926994.19   41.412546
0         Coal                     1150861.91   24.732883
4      Nuclear                      826910.84   17.770932
9         Wind                      318528.56    6.845417
3        Hydro                      300321.72    6.454138


Fuel Types and Total Energy Production Percentage for Europe:
  primary_fuel  estimated_generation_gwh_2017  Percentage
1          Gas                     1114

In [60]:
# Assuming df is your DataFrame

# Calculate the total energy produced by summing up the values from 'estimated_generation_gwh_2017'
total_energy_produced = df['estimated_generation_gwh_2017'].sum()

# Calculate the total CO2 emissions by summing up the values from 'Total MtCO2'
total_co2_emissions = df['Total MtCO2'].sum()

# Print the results
print("Total Energy Produced:", total_energy_produced)
print("Total CO2 Emissions:", total_co2_emissions)


Total Energy Produced: 23741500.669999998
Total CO2 Emissions: 15891269949.71


In [47]:
with pd.ExcelWriter('Co2_Emissions+Europe_with_groups.xlsx', engine='xlsxwriter') as writer:
    # Write each grouped dataframe to a different sheet
    total_energy_by_country.to_excel(writer, sheet_name='Total Energy by Country', index=False)
    energy_pollution_by_country_fuel.to_excel(writer, sheet_name='Energy Pollution Country Fuel', index=False)
    total_co2_by_country_df.to_excel(writer, sheet_name='Total CO2 by Country', index=False)
    
    # Write the original CSV data to a separate sheet
    df.to_excel(writer, sheet_name='Original Data', index=False)